[BE] Milestone API 기능 구현

suhyunsim opened the pull request:
* Milestone 목록, 생성, 수정, 삭제, picker api 구현
* picker api는 ModelMapper를 활용해서 구현한다.

suhyunsim committed:
[#33] feat: Milestone, MilestoneRepository 구현

Milestone 테이블과 연결할 Milestone 엔티티를 구현한다.
Milestone 목록을 가져오기 위해 MilestoneRepository를 구현한다.

suhyunsim committed:
[#33] feat: MilestoneService 구현

전체 Milestone 목록을 반환하는 findAll 메서드를 구현한다.
생성자 주입으로 MilestoneRepository를 주입한다.

suhyunsim committed:
[#33] feat: MilestoneController 구현

Milestone 전체 목록을 반환하도록 GET 매핑한다.

suhyunsim committed:
[#33] feat: Milestone 생성 기능 구현

save 메서드로 milestone을 생성 후 DB에 저장한다.
오류가 없을 시 201 Created Response를 반환한다.

suhyunsim committed:
[#33] feat: MilestoneController에 milestone을 생성하도록 PUT 매핑

suhyunsim committed:
[#33] feat: Unprocessable Entity 예외처리 구현

Milestone 생성 시 Request Body에 title이 null로 요청오는 경우를 예외처리한다.
ValidationException을 구현하고 해당 응답의 Http 상태 코드는 422로 한다.

suhyunsim committed:
[#33] refactor: milestone title이 null인 경우 예외처리

suhyunsim committed:
[#33] refactor: GlobalExceptionHandler에서 NotFoundException 추가

suhyunsim committed:
[#33] feat: Milestone 수정 기능 구현

findById로 수정하려는 milestone을 찾는다.
찾으려는 milestone id가 존재하지 않는다면 NotFoundException을 발생시킨다.
milestone 정보를 수정한다.

suhyunsim committed:
[#33] feat: MilestoneController에 수정 api PUT으로 매핑

suhyunsim committed:
[#33] feat: Milestone 삭제 기능 구현

suhyunsim committed:
[#33] feat: MilestoneController에 milestone을 삭제하도록 DELETE 매핑

존재하지 않는 milestone 삭제 요청 시 NotFoundException으로 예외처리

suhyunsim committed:
[#33] chore: modelMapper, junit, assertj 의존성 추가

suhyunsim committed:
[#33] feat: ModelMapper를 `@Bean`으로 등록

단위테스트를 위해 ModelMapper를 `@Bean`으로 등록한다.
Entity를 DTO에 매핑해주기 위해 ModelMapper 라이브러리를 이용한다.

suhyunsim committed:
[#33] feat: MilestonePickerDTO 구현

Milestone 엔티티에서 특정 필드만 담고 있는 MilestonePickerDTO를 구현한다.
Mock서버를 위해 만든 GetMilestoneListResponseDTO를 재사용한다.

suhyunsim committed:
[#33] feat: Milestone picker 메서드 구현

ModelMapper를 생성자 기반으로 주입한다.
Milestone 엔티티 목록 전체를 가져와서 modelMapper를 통해 DTO형태로 Object Mappping 한다.
PickerDTO의 전체 목록을 가져온다.

suhyunsim committed:
[#33] feat: MilestoneController에 picker 목록을 가져오는 GET Mapping 구현

blossun commented on BE/src/main/java/kr/codesquad/issuetracker09/domain/Milestone.java:
```diff
@@ -0,0 +1,32 @@
+    @Column(name = "due_on")
+    private LocalDate dueOn;
+
+    @Column(name = "number_of_open_issue")
+    private Integer numberOfOpenIssue;
+
+    @Column(name = "number_of_closed_issue")
+    private Integer numberOfClosedIssue;
```
열린/닫힌 이슈 갯수를 필드로 가지고 있는 것보다 메서드로 계산해서 가져오는 것은 어떨까요

suhyunsim replied:
엔티티에서 필드로 가지고 있던 부분을 삭제하고 계산해서 가져올 수 있도록 변경했습니다!

blossun commented on BE/src/main/java/kr/codesquad/issuetracker09/service/MilestoneService.java:
```diff
@@ -0,0 +1,58 @@
+
+    public Milestone save(Milestone milestone) {
+        if (milestone.getTitle() == null) {
+            throw new ValidationException("Title can't be blank");
```
Exception 처리 👍👍

blossun commented on BE/src/main/java/kr/codesquad/issuetracker09/service/MilestoneService.java:
```diff
@@ -0,0 +1,58 @@
+        List<Milestone> milestones = milestoneRepository.findAll();
+        List<MilestonePickerDTO> milestonePickerDTOList = new ArrayList<>();
+        for (Milestone milestone : milestones) {
+            MilestonePickerDTO milestonePickerDTO = modelMapper.map(milestone, MilestonePickerDTO.class);
```
modelMapper를 쓰니깐 코드중복이 없어서 너무 좋네요!! 저도 적용해서 리팩토링 할게요~~

blossun commented on BE/src/main/java/kr/codesquad/issuetracker09/web/milestone/controller/MilestoneController.java:
```diff
@@ -0,0 +1,69 @@
+        Optional<Milestone> originMileStone = milestoneService.findById(milestoneId);
+        if (!originMileStone.isPresent()) {
+            throw new NotFoundException("Milestone doesn't exist");
+        } else {
```
else 없이 바로 써도 좋을 것 같아요!

suhyunsim replied:
반영하겠습니다~

blossun reviewed (changes requested):
화이팅 고생 많으셨습니다!

suhyunsim committed:
[#33] feat: issue 엔티티 구현

milestone 엔티티와 매핑하기 위해 issue 엔티티를 구현한다.
다대일 양방향 관계로 매핑한다.

suhyunsim committed:
[#33] refactor: Milestone 엔티티 필드 수정

- Issue 엔티티와 Milestone 엔티티를 다대일 양방향 관계로 매핑한다.
- 열린/닫힌 issue를 계산하기 위해 get 메서드를 구현한다.
- 기존에 필드로 갖고 있던 이슈의 개수와 관련된 필드들을 삭제하고 계산해서 전달한다.

suhyunsim committed:
[#33] feat: MilestoneListDTO 구현

- 마일스톤 전체 목록을 전달하기 위한 DTO를 구현한다.
- 엔티티에서 계산할 열린/닫힌 이슈의 개수들도 전달한다.

suhyunsim committed:
[#33] refactor: Milestone 전체 목록을 엔티티가 아닌 DTO 형태로 전달

- 기존에 마일스톤 엔티티로 전체 목록을 전환하던 findAll 메서드를 MilestoneListDTO로 전달하도록 한다.
- modelMapper를 통해 엔티티의 형태를 DTO로 객체매핑 해준다.

suhyunsim committed:
[#33] refactor: milestone 목록 반환 메서드, 편집 메서드 수정

- getAllMilestones()의 반환 형태를 List<MilestoneListDTO>로 변경한다.
- 불필요한 else문을 삭제한다.

suhyunsim commented:
수정사항 반영했습니다!
mock 서버를 위해서 사용하던 GetMilestoneListResponseDTO를 재사용하려고 했는데 numberOfOpenIssue, numberOfClosedIssue를 계산하는 메서드의 반환형태가 Long이 되어버려서 사용하기가 어렵더라고요.
해당 사항만 다르게 반영한 DTO를 새로 생성해서 목록을 전달하도록 반영했습니다!
